contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -0,0 +1,216 @@
+   uint32 host_id = 2;
+   // size is the length of the mapping.
+   uint32 size = 3;
+}
+
+// UserNamespaceMapping is a user namespace mapping for uids and gids.
+message UserNamespaceMapping {
+   // uid_mappings is an array of user id mappings.
+   repeated LinuxIDMapping uid_mappings = 1;
+   // gid_mappings is an array of group id mappings.
+   repeated LinuxIDMapping gid_mappings = 2;
+}
+
+message UserNamespaceConfig {
+   // is_enabled, if true indicates that user-namespaces are supported and enabled in the container runtime
+   string is_enabled = 1;
+   // node_usernamespace_mapping is the user namespace mapping set for the container runtime
+   UserNSMapping node_usernamespace_mapping = 2;
+}
+message RuntimeConfig {
+    NetworkConfig network_config = 1;
+    UserNamespaceConfig user_namespace_config = 2;
+}
+
+message UpdateRuntimeConfigRequest {
+    RuntimeConfig runtime_config = 1;
+}
+
+message UpdateRuntimeConfigResponse {}
+
+message RuntimeConfigInfoRequest {}
+
+message RuntimeConfigInfoResponse {
+    RuntimeConfig runtime_config = 1
+}
+
+...
+
+// NamespaceOption provides options for Linux namespaces.
+message NamespaceOption {
+	// Network namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped network in the Kubernetes API.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceMode network = 1;
+	// PID namespace for this container/sandbox.
+	// Note: The CRI default is POD, but the v1.PodSpec default is CONTAINER.
+	// The kubelet's runtime manager will set this to CONTAINER explicitly for v1 pods.
+	// Namespaces currently set by the kubelet: POD, CONTAINER, NODE
+	NamespaceMode pid = 2;
+	// IPC namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped IPC in the Kubernetes API.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceMode ipc = 3;
+	// User namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped user namespace in the
+	// Kubernetes API. The container runtime has to run with user namespace enabled,
+	// and NODE then forces the use of host user namespace.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceModeValue user = 4;
+}
+
+// NamespaceModeValue is the wrapper of NamespaceMode, to allow non-POD default.
+message NamespaceModeValue {
+	NamespaceMode value = 1;
+}
+```
+
+### Runtime Support
+- Docker: https://docs.docker.com/engine/security/userns-remap/
+    - Concerns:
+Docker API does not provide user-namespace mapping. Therefore to handle `RuntimeConfigInfo` API, changes will be done in `dockershim` to read system files, `/etc/subuid` and `/etc/subgid`, for figuring out default user-namespace mapping. It won’t be possible to use custom user/group name to `--userns-remap parameter`, the implementation will expect `dockremap`, which is the default username.
+- CRI-O: WIP https://github.com/nalind/cri-o/tree/idmapping
+
```

Comment:
containerd has support for remapped snapshot.

https://github.com/containerd/containerd/blob/129167132c5e0dbd1b031badae201a432d1bd681/container_opts_unix.go#L149

Reply:
I thought the list in the document is intended to show the current implementation status of several CRI runtimes including containerd, but no strong opinion.

Reply:
@AkihiroSuda thanks for the pointer, the intention is to show the capability is supportable across a variety of runtime choices.  adding a section on containerd is helpful for everyone.

Reply:
:+1: We should add a line for containerd support as well.

Reply:
+1

/cc @Random-Liu

Reply:
@vikaschoudhary16, could we add a bullet point about the containerd work, to make the document more completely describe the landscape?